### DISCUSSION TITLE
Multiple docker containers in one workspace

### DISCUSSION THREAD

**Discussion by simonedelmann** [coder/coder#9067] [General] [answered]
Hello! This might be a dumb question, so I apologize beforehand. But I was looking for + trying things for more than an hour now and not getting anywhere:

I am trying to create a template for node/javascript development via the docker platform. I therefore modified the Dockerfile in the Docker starter template to have node/npm/etc. available. So far this is working as expected.

However, I need to add a Postgres container to the Terraform template. I tried to add the following block, but this one does not work at all, no matter what I tried. 

```
resource "docker_container" "database" {
  count = data.coder_workspace.me.start_count
  image = "postgres:alpine"
  name = "coder-${data.coder_workspace.me.owner}-${lower(data.coder_workspace.me.name)}-database"
  hostname = data.coder_workspace.me.name
  entrypoint = ["sh", "-c", replace(coder_agent.main.init_script, "/localhost|127\\.0\\.0\\.1/", "host.docker.internal")]
  env        = ["CODER_AGENT_TOKEN=${coder_agent.main.token}"]
  host {
    host = "host.docker.internal"
    ip   = "host-gateway"
  }
  volumes {
    container_path = "/var/lib/postgresql/data"
    volume_name    = docker_volume.home_volume.name
    read_only      = false
  }
}
```

Do you have any suggestion how I can create a Node+Postgres development template? 

Thank you very much for your help!

**Comment by uncvrd** [comment]
...did you ever figure this out? My current environment runs in a monorepo that uses docker compose for services like postgres and then turborepo for my typescript web/packages. I'd love to run both in a workspace but I have no idea how (devcontainers via envbuilder in Coder does not support docker-compose either which I thought would be my best shot)...the fact this project is called Coder makes it super difficult to search for solutions around the web too lol.

**Reply by simonedelmann** [reply]
Unfortunately no. I have stopped using Coder for now, because I couldn't figure this out and because I didn't really needed to use it.

**Reply by matifali** [reply]
Hi. Yes envbuilder doesn't support building from docker compose yet but there are a few workarounds. 

1. Setting up multiple docker containers in your workspace template. (I may add a sample template on how to achieve this give a docker compose file. @uncvrd it would be nice if you can share a sample devcontainer with docker compose you want to have as a docker workspace.)
2. Running a workspace with [`sysbox-runc`](https://coder.com/docs/templates/docker-in-workspaces#sysbox-container-runtime) will allow you to run docker within your workspace.

**Reply by ericpaulsen** [reply]
hi @uncvrd - adding to @matifali's 2nd point, here is our documentation on how you can run docker inside of a containerized workspace: https://coder.com/docs/templates/docker-in-workspaces

let us know if you have any questions and we're happy to help.

**Reply by uncvrd** [reply]
Hi @matifali & @ericpaulsen thank you both for the prompt feedback! Regardless of devcontainers, I think it it would be super helpful for me to share the current local setup and maybe provide how to best handle this scenario in a CDE? It's a lot and I'm struggling to figure out how to stitch it all together. Thanks in advance for any guidance.

Here is a simplified structure of the application. 

```
foundry/
├── compose-up.sh
├── docker-compose.yaml
├── services/
│   ├── typesense/
│   │   └── Dockerfile
│   ├── caddy/
│   │   └── Dockerfile
│   ├── vault/
│   │   └── Dockerfile
│   └── supabase/
│       └── config.toml
└── turborepo/
    ├── apps/
    │   └── web/
    │       └── package.json
    └── packages/
        └── ui/
            └── package.json
```

At a basic level there are two main subdirectories

1. services - a directory of (mostly) Docker images
2. turborepo - a monorepo directory of Typescript apps & shared packages/libs (using https://turbo.build/repo)

To start a development environment I must run two commands

1. `./compose-up.sh`

This starts the resources found in `/services`:

```
#!/bin/bash
docker-compose -f docker-compose.yaml up -d

(cd services/supabase && supabase start)
```

The second line is important because it starts a local version of Supabase (https://supabase.com/). Which means it spins up a separate collection of Docker containers (for postgres/admin ui/etc)...I assume I would need something like Docker-In-Docker for this part?

2. `cd turborepo && turbo dev`

This runs the `dev` script within each `package.json` in the `turborepo` directory

---

One of the `services` I deploy is `Caddy` (a reverse proxy that provisions TLS certs) so I can use https in local development. I have added `127.0.0.1 app.foundry-dev.ac` in my `etc/hosts` which means that I can visit `https://app.foundry-dev.ac` instead of `http://localhost:3000` which is proxied via `Caddy` to my `web` app.

With all that set up, I can start developing. As you can see, there's a lot of steps which is where I'm really excited to have Coder to help with.

I am not quite sure how I can translate this over in to a Coder VM and would ideally start running those two commands above so I can just visit `https://app.foundry-dev.ac` on my local device and start frontend development.

Let me know if I can provide any other information to help with moving things forward!

**Reply by matifali** [reply, marked as answer]
Hi @uncvrd Thanks for sharing the use case. I guess your best path forward is to use docker in workspaces. Please follow this guide to set up: https://coder.com/docs/templates/docker-in-workspaces#sysbox-container-runtime

After that you should be able to use your current workflow in the Coder workspace same as the local development.
